firewall/inspection/portscan/detect.go

```diff
@@ -0,0 +1,357 @@
+			entry.score = intMin(entry.score+addScore, maxScore)
+
+			if entry.previousOffender || entry.score >= scoreBlock {
+				entry.blocked = true
+				entry.previousOffender = true
+
+				//TODO: actually I just want to know if THIS threat exists - I don't need prefixing. Maybe we can do it simpler ... (less CPU-intensive)
+				if t, _ := status.GetThreats(threadPrefix + ipString); len(t) == 0 {
+					log.Tracer(ctx).Debugf("new Threat")
+					status.AddOrUpdateThreat(&status.Threat{
+						ID:              threadPrefix + ipString,
+						Name:            "Portscan by " + ipString,
+						Description:     "Someone tries to connect to a lot of closed Ports (non-running Services). Probably he wants to find out the services running on the maschine to determine which services to attack", //fixme: to long
+						MitigationLevel: status.SecurityLevelHigh,
+						Started:         time.Now().Unix(),
+					})
+				}
+			}
+		}
+
+		log.Tracer(ctx).Debugf("changed Entry: %+v", entry)
+	}
+}
+
+//updateScoreIgnoreBlockPrevOffender updates this 4 Values of the Struct
+//ipString needs to correspond to the key of the entry in the map ips
+//WARNING: This function maybe deletes the entry ipString from the Map ips. (look at the returncode)
+//return: still in map? (bool)
+func (d *ipData) updateScoreIgnoreBlockPrevOffender(ipString string) bool {
+	d.score -= intMin(int(time.Since(d.lastUpdated)/decreaseInterval), d.score)
+
+	if d.ignore {
+		if time.Since(d.lastSeen) > unignoreTime {
+			d.ignore = false
+		}
+	}
+
+	if d.previousOffender && time.Since(d.lastSeen) > undoSuspicionIdleTime {
+		d.previousOffender = false
+	}
+
+	if d.blocked && time.Since(d.lastSeen) > unblockIdleTime {
+		d.blocked = false
+		d.blockedPorts = []tcpUDPport{}
+
+		status.DeleteThreat(threadPrefix + ipString)
+	}
+
+	if !d.blocked && !d.ignore && !d.previousOffender && d.score == 0 {
+		delete(ips, ipString)
+		return false
+	}
+
+	d.lastUpdated = time.Now()
+	return true
+}
+
+// Destroy implements the destroy interface.
+func (d *Detector) Destroy() error {
+	return nil
+}
+
+// DetectorFactory is a primitive detection method that runs within the factory only.
+func DetectorFactory(conn *network.Connection, pkt packet.Packet) (network.Inspector, error) {
+	return &Detector{}, nil
+}
+
+// Register registers the encryption detection inspector with the inspection framework.
+func init() {
+	ips = make(map[string]*ipData)
+
+	//cleanup old Threads
+	threads, _ := status.GetThreats(threadPrefix)
+	for _, t := range threads {
+		status.DeleteThreat(t.ID)
+	}
+
+	module = modules.Register("portscan-detection", nil, start, nil, "base", "netenv")
+	module.Enable()
+}
+
+func updateWholeList() {
+	log.Debugf("Portscan detection: update list&cleanup")
+	for ip, entry := range ips {
+		//done inside the loop to give other goroutines time in between to access the list (and during that time block this task)
+		runOnlyOne.Lock()
+		defer runOnlyOne.Unlock()
```

Comment:
There's a deadlock here. The deferred unlock happens when the function returns but if there's more that on entry the second one will block on `runOnlyOne.Lock()`. This one only works for the very first interation and will deadlock afterwards (I remember you said shutdown is blocked and doesn't work?). 

How much entries do you expect to be in the `ips` map. IMHO it won't be a lot (likely always below 100, maybe even always below 10) so it's much faster to just update those few values than caring about locking, unlocking and notifying waiters of the mutex. Just lock at the beginning and unlock when done.

Reply:
Oh, you are right, defer isn't called each loop-run.